proof/refine/ARM_HYP/Arch_R.thy

```diff
@@ -2246,16 +2266,20 @@ lemma archThreadSet_ex_nonz_cap_to'[wp]:
   "archThreadSet f t \<lbrace>ex_nonz_cap_to' v\<rbrace>"
   unfolding ex_nonz_cap_to'_def cte_wp_at_ctes_of by wp
 
+definition associateVCPUTCB_helper where
+  "associateVCPUTCB_helper vcpu v t = do
+    archThreadSet (atcbVCPUPtr_update (\<lambda>_. Some v)) t;
+    setObject v (vcpuTCBPtr_update (\<lambda>_. Some t) vcpu)
+   od"
+
 lemma assoc_invs':
   "\<lbrace>invs' and
     ko_at' (vcpu\<lparr>vcpuTCBPtr:= None\<rparr>) v and
     obj_at' (\<lambda>tcb. atcbVCPUPtr (tcbArch tcb) = None) t and
     ex_nonz_cap_to' v and ex_nonz_cap_to' t\<rbrace>
-   do
-    archThreadSet (atcbVCPUPtr_update (\<lambda>_. Just v)) t;
-    setObject v $ vcpuTCBPtr_update (\<lambda>_. Just t) vcpu
-   od \<lbrace>\<lambda>_. invs'\<rbrace>"
-  unfolding invs'_def valid_state'_def valid_pspace'_def valid_mdb'_def
+   associateVCPUTCB_helper vcpu v t
```

Comment:
I'm really not clear how `associateVCPUTCB_helper` is helping compared to what we had before. Is there a reason you're doing it this way? Even if you want to force a forward proof, you could split off two statements with an assoc rule... here it looks like you're collapsing two statements into a def, making a rewrite rule for those two lines to fold them into the def, then proving a wp rule over a def where it used to be two lines, and then doing a symmetric subst for the def later on... hard to follow/understand why.

Reply:
(where you = either actual you or whoever actually did this)

Reply:
This was very annoying. I was able to easily do forward reasoning the whole way through, but then I would have been specifying preconditions that would have come out of the wp machinery... So I tried to follow the previous proof.

The problem is that this bundling (that now occurs in this helper function) is now not right towards the end of the function, now that `vcpuSwitch` is there. And the rewrite rule that was there before wouldn't apply nicely after doing some `wpsimp`ing to take care of the `vcpuSwitch`. 

I think I can get rid of the helper function and the rewrite rule with a forwards proof. Would that be the best way of going?

Reply:
You can do something like what you're doing now, but with less repetition and more constraint to within the single proof you're using it in (unless you're using it in more proofs that I missed?). At the very least, the definition and rewrite (fold) are redundant, because you can already pick a definition that folds the way you want.
This might also be an opportunity to practice Isar-style proofs using proof-internal definitions.

Reply:
I tried for hours and hours to get it working without the helper definition, but I think the associativity of bind was killing me: there didn't seem to be a nice way to have a function with many many binds, and then separate precisely the last two of them. Any ideas?